pkg/blockbuilder/blockbuilder.go

```diff
@@ -239,3 +239,298 @@ func (b *BlockBuilder) getLagForPartition(ctx context.Context, partition int32)
 
 	return kadm.GroupMemberLag{}, lastErr
 }
+
+type partitionState struct {
+	// Partition is the partition the state belongs to.
+	Partition int32
+	// Lag is the upperbound number of records the cycle will need to consume.
+	Lag int64
+	// Commit is the current offset commit for the partition.
+	Commit kadm.Offset
+	// CommitRecTimestamp is the timestamp of the record whose offset was committed (and not the time of commit).
+	CommitRecTimestamp time.Time
+	// OldestSeenOffset is the offset of the last record consumed in the commiter-cycle. It can be greater than Commit.Offset if previous cycle overconumed.
+	OldestSeenOffset int64
+	// LastBlockEnd is the timestamp of the block end in the commiter-cycle.
+	LastBlockEnd time.Time
+}
+
+func partitionStateFromLag(logger log.Logger, lag kadm.GroupMemberLag, fallbackMillis int64) *partitionState {
+	commitRecTs, oldestSeenOffset, lastBlockEndTs, err := unmarshallCommitMeta(lag.Commit.Metadata)
+	if err != nil {
+		// If there is an error in unmarshalling the metadata, treat it as if
+		// we have no commit. There is no reason to stop the cycle for this.
+		level.Warn(logger).Log("msg", "error unmarshalling commit metadata", "err", err, "partition", lag.Partition, "offset", lag.Commit.At, "metadata", lag.Commit.Metadata)
+	}
+
+	if commitRecTs == 0 {
+		// If there was no commit metadata, we use the fallback to replay a set amount of
+		// records because it is non-trivial to peek at the first record in a partition to determine
+		// the range of replay required. Without knowing the range, we might end up trying to consume
+		// a lot of records in a single partition consumption call and end up in an OOM loop.
+		// TODO(codesome): add a test for this.
+		commitRecTs = fallbackMillis
+	}
+
+	return &partitionState{
+		Partition:          lag.Partition,
+		Lag:                lag.Lag,
+		Commit:             lag.Commit,
+		CommitRecTimestamp: time.UnixMilli(commitRecTs),
+		OldestSeenOffset:   oldestSeenOffset,
+		LastBlockEnd:       time.UnixMilli(lastBlockEndTs),
+	}
+}
+
+func (b *BlockBuilder) consumePartition(ctx context.Context, state *partitionState, cycleEnd time.Time) error {
+	sectionCycleEnd := cycleEnd
+	if sectionCycleEnd.Sub(state.CommitRecTimestamp) > 3*b.cfg.ConsumeInterval/2 {
+		// We are lagging behind by more thatn 1.5*interval or there is no commit. We need to consume the partition in sections.
```

Comment:
```suggestion
// We are lagging behind by more than 1.5*interval or there is no commit. We need to consume the partition in sections.
```

Reply:
leaving suggestions from goland is a mess. i meant this 
```suggestion
		// We are lagging behind by more than 1.5*interval or there is no commit. We need to consume the partition in sections.
```